### PR TITLE
[CI] Upgrade GoogleTest version from 1.12.1 to 1.13.0

### PR DESCRIPTION
**Pull request by cngzhnp**
Upgraded GoogleTest to `1.13.0` + `C++14`

## Changes

Revised the versions used for GoogleTest:

* By default, upgraded to `1.13.0`, which requires `C++14`
* Still using `1.12.1` (unchanged) for CI builds that requires `C++11`
* Still using `1.10.0` (unchanged) for CI builds using `GCC 4.8`

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: ci/setup_cmake.sh
Comment (lalitb) [inline ci/setup_cmake.sh:19]: Why is default version changed to 1.10.0, can it remain as 1.12.1 ?

**Comment by lalitb** [comment]
Can you update the description for the changes. As the latest version doesn't support C++11, how are we handling now CI tests for C++11 ?

File: ci/setup_cmake.sh
Comment (cngzhnp) [reply ci/setup_cmake.sh:19]: If GoogleTest version is not set on CI Pipeline, it was set as 1.10.0 by default. It could not be 1.12.1 version because GCC 4.8 oldish environment to support C++14.

File: CMakeLists.txt
Comment (lalitb) [inline CMakeLists.txt:538]: @cngzhnp  - I think this will break the builds for users who are building library with tests using gcc4.8 ( or any C++11 only compiler) ?

Comment (cngzhnp) [reply CMakeLists.txt:538]: Yes, you are right. However, it is like CI/CD pipeline design problem for opentelemetry-cpp. There are a lot of scripts which setup different versions of third party libraries or submodules etc. All of them are not in one page.

From my perspective, if we still want to support GCC 4.8 version with C++11 standard, we need a better CMake/CI-CD/Script design for this. Otherwise, like me people might only focus on passing the CI tests but not the real usecases.

File: .github/workflows/ci.yml
Comment (ThomsonTan) [inline .github/workflows/ci.yml:280]: Why is this version requirement for GCC 4.8 removed?

Comment (cngzhnp) [reply .github/workflows/ci.yml:280]: It is not removed, it was set in the build script. As I wrote above comments, "If GoogleTest version is not set on CI Pipeline, it was set as 1.10.0 by default. It could not be 1.12.1 version because GCC 4.8 oldish environment to support C++14."

File: CHANGELOG.md
Comment (marcalff) [inline CHANGELOG.md:62]: These lines are not added by this PR, and it should not show here.

There is something wrong the way merged from the main branch to this PR branch are done.

File: .github/workflows/ci.yml
Comment (marcalff) [reply .github/workflows/ci.yml:280]: The logic is not intuitive, and convoluted in my opinion.

Instead of:

* setting a very old version by default in setup_cmake
* overriding with a new version in ci.yml in every places expect a few old builds

I suggest the opposite:

* Set by default the most up to date version in setup_cmake (1.13.0)
* Override the version in ci.yml for builds where we do want, on purpose, an older version (gcc 4.8, and the new C++ build to come in PR #2152)

File: ci/do_ci.sh
Comment (marcalff) [inline ci/do_ci.sh:199]: CMAKE_CXX_STANDARD=20 is valid for cmake.c++20.stl.test, ok.

For all the builds that now uses Gtest 1.13.0, which require C++14, ci/do_ci.sh most likely needs to be updated to add CMAKE_CXX_STANDARD=14 in relevant builds.

File: CMakeLists.txt
Comment (marcalff) [reply CMakeLists.txt:543]: This sets C++20 for all builds using tests, which breaks gcc4.8.

See related comment, to set C++14 is relevant ci/do_ci.sh targets instead.

**Review by marcalff — COMMENTED** [review]
Thanks for the PR.

See comments for action items.

**Review by marcalff — COMMENTED** [review]
@cngzhnp 

Now that the SDK header file cleanup is merged (thanks for the PR),
it is now time to look again at GTest.

First, please note I did a main -> PR 2114 merge to get up to date code.

As a result, be sure to pull before making any new changes.

Secondly, there is now a new build in CI:

CI / CMake gcc 12 (maintainer mode, C++11, async) (pull_request) 

It currently fails, so please fix it by using GTest version 1.12.1 and C++11.

The ci/setup_cmake.sh file needs to be fixed as follows:

When GOOGLETEST_VERSION is empty, the default value should be the most up to date version, so 1.13.0

Then, when proceeding to download GTest, the script must check the version number to build the proper GOOGLETEST_VERSION_PATH / GOOGLETEST_FOLDER_PATH.

This must work for 1.10.0, 1.12.1 and 1.13.0, and is independent of whether a value was provided by the caller or not.

Last, please cleanup the extra entry in the CHANGELOG.

**Comment by marcalff** [comment]
Ping @cngzhnp

Did you see the previous comments ?

**Comment by cngzhnp** [comment]
Hello @marcalff, sorry for the late answer. Yeah, I saw that but I was busy with other stuff. I planned to work this task on this week.

**Comment by cngzhnp** [comment]
@marcalff Could you please verify the changes after your comments? Also, could you please take a look that why CodeQL build is failed, I am not sure that fail is directly related to my changes?

**Comment by marcalff** [comment]
@cngzhnp Thanks for the cleanup.

Remaining points below.

## CodeQL

The `CodeQL / CodeQL-Build` build fails on:

```
  [  0%] Building CXX object api/test/core/CMakeFiles/version_test.dir/version_test.cc.o
  cd /home/runner/work/opentelemetry-cpp/opentelemetry-cpp/_lgtm_build_dir/api/test/core && /usr/bin/c++  -DENABLE_TEST -I/home/runner/work/opentelemetry-cpp/opentelemetry-cpp/api/include  -fpermissive -O3 -DNDEBUG   -Wno-error=deprecated-declarations -std=gnu++11 -o CMakeFiles/version_test.dir/version_test.cc.o -c /home/runner/work/opentelemetry-cpp/opentelemetry-cpp/api/test/core/version_test.cc
  In file included from /usr/include/gtest/gtest-message.h:57,
                   from /usr/include/gtest/gtest-assertion-result.h:46,
                   from /usr/include/gtest/gtest.h:64,
                   from /home/runner/work/opentelemetry-cpp/opentelemetry-cpp/api/test/core/version_test.cc:6:
  /usr/include/gtest/internal/gtest-port.h:270:2: error: #error C++ versions less than C++14 are not supported.
    270 | #error C++ versions less than C++14 are not supported.
        |  ^~~~~
```

To fix this and keep the build in C++11, add:

```
GOOGLETEST_VERSION: 1.12.1
```

in file codeql-analysis.yml

## setup-cmake.sh

Per previous comments, this code:

```
# This variable not set on CI pipeline for only legacy environment(GCC 4.8).
# With 1.13.0 version, C++14 must be set which does not supported by legacy environment anymore.
# Also with this version, release version path needs to be adapted.
if [ -z "${GOOGLETEST_VERSION}" ]; then
  # By default, GoogleTest version set the following version.
  export GOOGLETEST_VERSION=1.13.0
  GOOGLETEST_VERSION_PATH="v${GOOGLETEST_VERSION}"
  GOOGLETEST_FOLDER_PATH="googletest-${GOOGLETEST_VERSION}"
else
  GOOGLETEST_VERSION_PATH="release-${GOOGLETEST_VERSION}"
  GOOGLETEST_FOLDER_PATH="googletest-release-${GOOGLETEST_VERSION}"

fi
```

should decouple setting the default release with the release format.

Something like:

```
if [ -z "${GOOGLETEST_VERSION}" ]; then
  # Version by default. Requires C++14
  export GOOGLETEST_VERSION=1.13.0
fi
```

followed by something like:

```
if [  << GOOGLETEST_VERSION greater or equal to 1.13.0 >> ]; then
  GOOGLETEST_VERSION_PATH="v${GOOGLETEST_VERSION}"
  GOOGLETEST_FOLDER_PATH="googletest-${GOOGLETEST_VERSION}"
else
  GOOGLETEST_VERSION_PATH="release-${GOOGLETEST_VERSION}"
  GOOGLETEST_FOLDER_PATH="googletest-release-${GOOGLETEST_VERSION}"
fi
```

Expectation is that all the following use case works

```
unset GOOGLETEST_VERSION
setup_cmake
```

installs 1.13.0

```
GOOGLETEST_VERSION=1.13.0
setup_cmake
```

installs 1.13.0

```
GOOGLETEST_VERSION=1.12.1
setup_cmake
```

installs 1.12.1

```
GOOGLETEST_VERSION=1.10.0
setup_cmake
```

installs 1.10.0

**Comment by cngzhnp** [comment]
@marcalff Could you please check & verify what I changed so for? Now, I fixed the CodeQL and split up GoogleTest version and its path.

**Comment by marcalff** [comment]
> @marcalff Could you please check & verify what I changed so for? Now, I fixed the CodeQL and split up GoogleTest version and its path.

Thanks.

How about this instead:

```
OLD_VERSION_REGEXP="^1\.([0-9]|10|11|12)(\..*)?$"

if [[ ${GOOGLETEST_VERSION} =~ ${OLD_VERSION_REGEXP} ]]; then
  # Old package name
else
  # new (starting with 1.13.0) package name
fi
```

File: .github/workflows/ci.yml
Comment (marcalff) [reply .github/workflows/ci.yml:311]: Marking as resolved, the version 1.10.0 is back in the CI script

**Review by marcalff — APPROVED** [review]
The code LGTM.

Please revise:

* the PR description
* code comments

for cleanup.

@owent 

This change may cause merge collisions with your work, when setting CMAKE_CXX_STANDARD. I assume these will be trivial to fix, so not an issue.

@lalitb @esigo @owent 

Please take a second look.

**Review by owent — APPROVED** [review]
LGTM and thanks.

**Comment by owent** [comment]
> The code LGTM.
> 
> Please revise:
> 
> * the PR description
> * code comments
> 
> for cleanup.
> 
> @owent
> 
> This change may cause merge collisions with your work, when setting CMAKE_CXX_STANDARD. I assume these will be trivial to fix, so not an issue.
> 
> @lalitb @esigo @owent
> 
> Please take a second look.

It's fine by me , we also use gtest 13 and a high `CMAKE_CXX_STANDARD` in some of our products.While for the users still use gcc 4.8 and C++11, them can still use gtest 1.10.0 and it works well now.

**Comment by marcalff** [comment]
Set merge temporarily on hold, to avoid conflicts with the release of opentelemetry-cpp 1.10.0.

File: CMakeLists.txt
Comment (ThomsonTan) [inline CMakeLists.txt:212]: Could this option be applied to test code only instead of making it a global option?

File: ci/do_ci.sh
Comment (lalitb) [inline ci/do_ci.sh:92]: This property would be redundant once we switch to C++11 (#2147), as we would update the property [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/ca8c23481ae10c1569858cee19fa076c27b8f261/CMakeLists.txt#L139). Need to make a note of removing from here once that is done.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for the changes.

File: ci/do_ci.sh
Comment (marcalff) [reply ci/do_ci.sh:92]: Added a note in #2147, so this part is resolved (nothing to change in this PR).

File: CMakeLists.txt
Comment (marcalff) [reply CMakeLists.txt:220]: As discussed in C++ SIG meeting, 2023-07-12, this is a minor item that will be addressed in a separate PR.

**Pull request merged**
